### PR TITLE
fix(adapter-cloudflare): Generate folder path with wildcard if folder exists

### LINKED ISSUE #9616: adapter-cloudflare generates very inefficient _routes.json which hits rules limits
### Describe the bug

`@sveltejs/adapter-cloudflare` generates a quite inefficient `_routes.json` while cloudflare pages can only handle up to 100 rules. This ends up with this kind of warning:
```
Function includes/excludes exceeds _routes.json limits (see https://developers.cloudflare.com/pages/platform/functions/routing/#limits). Dropping 67 exclude rules — this will cause unnecessary function invocations.
```
For instance, the below directory structure:
```
static/images/{a,b,c,d,e}.png
static/fonts/{a,b,c,d,e}.otf
...
```
will list individual files and result into the below `_routes.json` exclude
```
images/a.png
...
images/e.png
fonts/a.otf
...
fonts/e.otf
```

while you'd expect some wildcards instead
```
images/*
fonts/*
```

### Reproduction

add lots of files to a folder `static/folder1`

### Logs

_No response_

### System Info

```Shell
System:
    OS: Windows 10 10.0.22621
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
    Memory: 14.95 GB / 31.80 GB
  Binaries:
    Node: 16.16.0 - ~\AppData\Local\pnpm\node.EXE
    Yarn: 1.22.19 - ~\AppData\Local\pnpm\yarn.CMD
    npm: 8.11.0 - ~\AppData\Local\pnpm\npm.CMD
  Browsers:
    Chrome: 111.0.5563.147
    Edge: Spartan (44.22621.1485.0), Chromium (111.0.1661.62)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @sveltejs/adapter-auto: 2.0.0 => 2.0.0
    @sveltejs/adapter-cloudflare: 2.2.0 => 2.2.0
    @sveltejs/kit: 1.15.2 => 1.15.2
    svelte: ^3.58.0 => 3.58.0
    vite: ^4.2.1 => 4.2.1
```

### Severity

annoyance

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by u-yas**
fixes #9616

The auto-generated _routes.json will output all files nested in the exclude array. This quickly exceeds the cloudflare limit of 100 rules.
Therefore, when it is determined that a folder is included in the output file, it has been modified to output the shallowest folder and wildcard

example (my private sveltekit project using adapter-cloudflare)
```
❯  tree static 
static
├── aaaaaads.txt
├── blog_ogp.png
├── favicon.ico
├── favicons
│   ├── a
│   │   ├── android-chrome-192x192.png
│   │   └── b
│   ├── apple-touch-icon.png
│   ├── browserconfig.xml
│   ├── c
│   │   └── android-chrome-384x384.png
│   ├── favicon-32x32.png
│   ├── favicon.ico
│   ├── favicon.png
│   └── mstile-150x150.png
├── ogp.png
├── robots.txt
├── site.webmanifest
└── x
    ├── safari-pinned-tab.svg
    └── y
```

output:
```json
{
	"version": 1,
	"description": "Generated by @sveltejs/adapter-cloudflare",
	"include": [
		"/*"
	],
	"exclude": [
		"/_app/*",
		"/aaaaaads.txt",
		"/blog_ogp.png",
		"/favicon.ico",
		"/favicons/*",
		"/ogp.png",
		"/robots.txt",
		"/site.webmanifest",
		"/x/*",
		"/about"
	]
}
```

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: packages/adapter-cloudflare/index.js [hunk 1]
```diff
@@ -99,7 +99,20 @@ function get_routes_json(builder, assets, { include = ['/*'], exclude = ['<all>'
 								file === '_redirects'
 							)
 					)
-					.map((file) => `/${file}`);
+					.reduce((prev, file_path) => {
+						const split_path = file_path.split('/');
+						/**
+						 * There is a limit of 100 rules or less. Take advantage of wildcards and generate as few file paths as possible
+						 * https://developers.cloudflare.com/pages/platform/functions/routing/#create-a-_routesjson-file
+						 */
```
Comment (benmccann) [inline packages/adapter-cloudflare/index.js:107]: should use `//` comment style here
```suggestion
						// There's a limit of 100 rules. Take advantage of wildcards and generate as few file paths as possible
						// https://developers.cloudflare.com/pages/platform/functions/routing/#create-a-_routesjson-file
```

Comment (u-yas) [reply packages/adapter-cloudflare/index.js]: Thank you!
Updated!!

**Comment by u-yas** [comment]
@benmccann 
Hello. Thank you for your maintenance.

Is there anything missing in re-reviewing this PR? If you are simply busy and can't see it, I will wait patiently.

**Review by ghostdevv — APPROVED** [review]

**Comment by ghostdevv** [comment]
It's a shame there are no tests as this would be a great thing to test

**Comment by u-yas** [comment]
> It's a shame there are no tests for this adapter as this would be a great thing to test

It's hard to test ...
Take some snapshots. Or, try to connect Cloudflare Pages to Github like this repository does with vercel to see if it can actually Deploy.... That's about all I can think of...

File: packages/adapter-cloudflare/index.js [hunk 2]
```diff
@@ -99,7 +99,17 @@ function get_routes_json(builder, assets, { include = ['/*'], exclude = ['<all>'
...
 							)
 					)
-					.map((file) => `/${file}`);
+					.reduce((prev, file_path) => {
+						const split_path = file_path.split('/');
+						// There's a limit of 100 rules. Take advantage of wildcards and generate as few file paths as possible
+						// https://developers.cloudflare.com/pages/platform/functions/routing/#create-a-_routesjson-file
+						if (split_path.length > 1) {
+							const exclude_pattern = `/${split_path[0]}/*`;
+
```
Comment (benmccann) [inline packages/adapter-cloudflare/index.js:108]:
```suggestion
							const exclude_pattern = `/${split_path[0]}/*`;
```

Comment (benmccann) [inline packages/adapter-cloudflare/index.js:108]:
```suggestion
```

**Review by benmccann — CHANGES_REQUESTED** [review]
I don't think this change is safe. E.g. in the example you gave in the PR description, how do you know there aren't any routes that would overlap with `"/x/*"`?

**Comment by u-yas** [comment]
Your point is correct...

Ideally, it matches sveltekit's page routing and avoids wildcards when it matches. I think

**Comment by u-yas** [comment]
updated
Utilize the regular expression in the pattern property contained in each element of the build.routes array and check

File: packages/adapter-cloudflare/index.js [hunk 3]
```diff
@@ -63,6 +62,35 @@ export default function (options = {}) {
...
+ * @param {string} file_path
+ */
+function find_shallowest_exclude_pattern(builder, file_path) {
+	const exclude_file_path = `/${file_path}`;
+	const path_segments = file_path.split('/');
+	if (path_segments.length === 0) {
+		return file_path;
+	}
+	let exclude_pattern = exclude_file_path;
+	// Exclude patterns can conflict with sveltekit routes, so don't wildcard paths in that case.
```
Comment (benmccann) [inline packages/adapter-cloudflare/index.js:79]: from a quick glance at this, I can't tell exactly what the logic is that's being used, but I would guess it's not correct. E.g. I think you would need to consider whether a route segment has a matcher and it doesn't look like you're checking for that. https://kit.svelte.dev/docs/advanced-routing#matching

Comment (u-yas) [reply packages/adapter-cloudflare/index.js:79]: I see, sveltekit has a lot of matching patterns.  informative.  I'm starting to think maybe I should add some tests before I publish a PR for this feature.  I would like to create a pr with reference to other packages.  On top of that, I will - work on this pr again 🙏

Thank you for taking time out of your busy schedule to review my PR.  I will make this PR a draft for now.

**Comment by pauldemarco** [comment]
Ran into needing this. I'd like to include `<prerendered>` in the excludes list but it shouldn't generate any individual routes if they are excluded from another wildcard entry `/api/comments/*`.

Is this how the PR works?

**Pull request closed without merging**